File: facebook/p85/Solution.java
/**
 * Given a 2D binary matrix filled with 0's and 1's,
 * find the largest rectangle containing only 1's and return its area.
 */

/**
 *
 * use the similar method as p84 largest area in histogram
 */

package facebook.p85;

import java.util.Stack;

public class Solution {
  public int maximalRectangle(char[][] matrix) {
    if(matrix.length == 0 || matrix[0].length == 0) return 0;

    int maxArea = 0;
    int currentArea = 0;
    int numOfRows = matrix.length;
    int numOfCols = matrix[0].length;
    // this is a trick. we use numOfCols + 1 as the length of the histogram heights
    // the last height is always 0, which is the lowest height
    int[] heights = new int[numOfCols + 1];

    for(int row = 0; row < numOfRows; row++) {
      Stack<Integer> stack = new Stack<>();

      for(int i = 0; i < numOfCols + 1; i++) {
        // construct the heights histogram
        if(i < numOfCols) {
          if(matrix[row][i] == '1') {
            heights[i] += 1;
          } else {
            heights[i] = 0;
          }
        }

        while(!stack.isEmpty() && heights[i] < heights[stack.peek()]) {
          int minHeightIdx = stack.pop();

          if(stack.isEmpty()) {
            currentArea = heights[minHeightIdx] * i;
          } else {
            // width = the distance between i and stack.peek()
            // i - stack.peek() - 1 because we should the count the 1 width of column i
            currentArea = heights[minHeightIdx] * (i - stack.peek() - 1);
          }

          maxArea = Math.max(maxArea, currentArea);
        }

        stack.push(i);
      }
    }

    return maxArea;
  }
}

File: facebook/p85/Solution.py
class Solution:
    def maximalRectangle(self, matrix):
        """
        :type matrix: List[List[str]]
        :rtype: int
        """
        if not matrix or not matrix[0]:
            return 0

        max_rectangle = 0
        num_of_rows = len(matrix)
        num_of_cols = len(matrix[0])
        heights = [0 for _ in range(num_of_cols)]

        for i in range(num_of_rows):
            for j in range(num_of_cols):
                if matrix[i][j] == '1':
                    heights[j] += 1
                else:
                    heights[j] = 0
            current_rectangle = self.max_hist(heights)
            print(heights)
            print(current_rectangle)
            max_rectangle = max(max_rectangle, current_rectangle)

        return max_rectangle

    def max_hist(self, heights):
        asc_height_idx_stack = []
        max_area = 0
        size = len(heights)

        for i in range(size + 1):
            current = heights[i] if i < size else -1
            while asc_height_idx_stack and heights[asc_height_idx_stack[-1]] >= current:
                top = asc_height_idx_stack.pop()

                if asc_height_idx_stack:
                    current_area = heights[top] * (i - asc_height_idx_stack[-1] - 1)
                else:
                    current_area = heights[top] * i
                print(current_area)
                max_area = max(max_area, current_area)

            asc_height_idx_stack.append(i)
        return max_area


def main():
    test = [
        ["1", "0", "1", "0", "0"],
        ["1", "0", "1", "1", "1"],
        ["1", "1", "1", "1", "1"],
        ["1", "0", "0", "1", "0"]
    ]

    sol = Solution()
    print(sol.maximalRectangle(test))

if __name__ == '__main__':
    main()
